### PR TITLE
Mygcpsteps

### PR DESCRIPTION
**Pull request by artvalborgcp**
Create and destroy infrastructure with terraform

### CODE REVIEW

File: 7-gce/bookshelf/tasks.py [hunk 1]
```diff
@@ -16,13 +16,13 @@
 
 from bookshelf import get_model, storage
 from flask import current_app
-from google.cloud import pubsub
+from google.cloud import pubsub_v1
```
Comment (yevheniimykhailin) [inline 7-gce/bookshelf/tasks.py]: from google.cloud import pubsub_v1 as pubsub

File: 7-gce/config.py [hunk 2]
```diff
@@ -32,7 +32,7 @@
 
 # Google Cloud Project ID. This can be found on the 'Overview' page at
 # https://console.developers.google.com
-PROJECT_ID = 'mygcpproject-314813'
+PROJECT_ID = 'my-gcp-terraform'
```
Comment (yevheniimykhailin) [inline 7-gce/config.py]: import os

PROJECT_ID = os.environ.get(PROJECT_ID, 'my-gcp-terraform')

you can pass via env vars

File: 7-gce/config.py [hunk 3]
```diff
@@ -43,7 +43,7 @@
 # Set this value to the Cloud SQL connection name, e.g.
 #   "project:region:cloudsql-instance".
 # You must also update the value in app.yaml.
-CLOUDSQL_CONNECTION_NAME = 'mygcpproject-314813:us-central1:bookshelf'
+CLOUDSQL_CONNECTION_NAME = 'my-gcp-terraform:us-central1:bookshelf'
```
Comment (yevheniimykhailin) [inline 7-gce/config.py]: same here

File: 7-gce/config.py [hunk 4]
```diff
@@ -90,7 +90,7 @@
 #
 # You can adjust the max content length and allow extensions settings to allow
 # larger or more varied file types if desired.
-CLOUD_STORAGE_BUCKET = 'your-bucket-name'
+CLOUD_STORAGE_BUCKET = 'my-gcp-terraform-bucket'
```
Comment (yevheniimykhailin) [inline 7-gce/config.py]: same here

File: terraform/instance.tf [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+  source_instance_template = google_compute_instance_template.tpl.id
+  network_interface {
+    network    = "mygcpnet"
+    subnetwork = "mygcpsubnet"
+    access_config {
+    }
+  }
+  depends_on = [google_compute_subnetwork.mygcpsubnet]
+}
+*/
```
Comment (yevheniimykhailin) [inline terraform/instance.tf]: remove if not needed

File: terraform/instance.tf [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+resource "google_compute_health_check" "autohealing" {
+  name                = "${var.name_prefix}-health-check"
+  project             = var.project_id
+  check_interval_sec  = 5
+  timeout_sec         = 5
+  healthy_threshold   = 2
+  unhealthy_threshold = 10 # 50 seconds
```
Comment (yevheniimykhailin) [inline terraform/instance.tf]: as variables with default values

File: terraform/mynetwork.tf [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+
+    path_rule {
+      paths   = ["/*"]
+      service = google_compute_backend_service.default.id
+    }
+  }
+}
+
+resource "google_compute_target_http_proxy" "http" {
+  count   = "1"
```
Comment (yevheniimykhailin) [inline terraform/mynetwork.tf]: Why do you put count=1?

File: terraform/outputs.tf [hunk 8]
```diff
@@ -0,0 +1,13 @@
+/*output "bookshelfinstance_ip" {
+  value = google_compute_instance_from_template.tpl.network_interface.0.network_ip
+}
+
+output "bookshelfinstance_id" {
+  value = google_compute_instance_from_template.tpl.id
+}
+*/
+
```
Comment (yevheniimykhailin) [inline terraform/outputs.tf:10]: Add more outputs

File: terraform/sql.tf [hunk 9]
```diff
@@ -0,0 +1,45 @@
+
+resource "google_compute_global_address" "mygcpprivate_ip" {
+  provider = google-beta
+
```
Comment (yevheniimykhailin) [inline terraform/sql.tf:2]: add labels

File: terraform/sql.tf [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+resource "google_sql_database_instance" "instance" {
+  provider = google-beta
+
+  name   = "bookshelf-${random_id.db_name_suffix.hex}"
+  region = "us-central1"
+
+  depends_on = [google_service_networking_connection.mygcp_vpc_connection]
+
+  settings {
+    tier = "db-f1-micro"
```
Comment (yevheniimykhailin) [inline terraform/sql.tf]: as variable

File: terraform/sql.tf [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+
+resource "random_id" "db_name_suffix" {
+  byte_length = 4
+}
+
+resource "google_sql_database_instance" "instance" {
+  provider = google-beta
+
+  name   = "bookshelf-${random_id.db_name_suffix.hex}"
+  region = "us-central1"
```
Comment (yevheniimykhailin) [inline terraform/sql.tf]: as variable

File: terraform/sql.tf [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+
+  depends_on = [google_service_networking_connection.mygcp_vpc_connection]
+
+  settings {
+    tier = "db-f1-micro"
+    ip_configuration {
+      ipv4_enabled    = true
```
Comment (yevheniimykhailin) [inline terraform/sql.tf]: external IP?

File: terraform/variables.tf [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+  type        = string
+  default     = "1"
+}
+variable "url_map" {
+  description = "The number of running instances"
+  type        = string
+  default     = "/_ah/health"
+}
+variable "service_port" {
+  description = "Port the service is listening on."
```
Comment (yevheniimykhailin) [inline terraform/variables.tf:181]: add type

File: terraform/variables.tf [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+variable "service_port_name" {
+  description = "Name of the port the service is listening on."
+}
```
Comment (yevheniimykhailin) [inline terraform/variables.tf:268]: add type

File: terraform/variables.tf [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+
+variable "auto_delete" {
+  description = "Whether or not the boot disk should be auto-deleted"
+  default     = "true"
+}
+
+variable "metadata_startup_script" {
+  description = "User startup script to run when instances spin up"
+  default     = ""
```
Comment (yevheniimykhailin) [inline terraform/variables.tf]: add type

File: terraform/variables.tf [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+variable "disk_type" {
+  description = "Boot disk type, can be either pd-ssd, local-ssd, or pd-standard"
+  default     = "pd-standard"
+}
+
+variable "auto_delete" {
+  description = "Whether or not the boot disk should be auto-deleted"
+  default     = "true"
```
Comment (yevheniimykhailin) [inline terraform/variables.tf]: should be bool with true without ""

File: terraform/variables.tf [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+  default     = null
+}
+variable "source_image" {
+  description = "Source disk image. If neither source_image nor source_image_family is specified, defaults to the latest public CentOS image."
+  default     = ""
+}
+
+variable "disk_size_gb" {
+  description = "Boot disk size in GB"
+  default     = "10"
```
Comment (yevheniimykhailin) [inline terraform/variables.tf]: should be number with just 10

File: terraform/variables.tf [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+variable "can_ip_forward" {
+  description = "Enable IP forwarding, for NAT instances for example"
+  default     = "false"
```
Comment (yevheniimykhailin) [inline terraform/variables.tf]: bool

File: terraform/variables.tf [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+variable "disk_type" {
+  description = "Boot disk type, can be either pd-ssd, local-ssd, or pd-standard"
+  default     = "pd-standard"
```
Comment (yevheniimykhailin) [inline terraform/variables.tf]: type

**Review by yevheniimykhailin — CHANGES_REQUESTED** [review]

File: 7-gce/config.py (hunk 2 above)
Comment (artvalborgcp) [reply 7-gce/config.py]: fixed

File: 7-gce/config.py (hunk 4 above)
Comment (artvalborgcp) [reply 7-gce/config.py]: fixed

File: terraform/instance.tf (hunk 5 above)
Comment (artvalborgcp) [reply terraform/instance.tf]: removed

File: terraform/instance.tf (hunk 6 above)
Comment (artvalborgcp) [reply terraform/instance.tf]: complete

File: terraform/mynetwork.tf (hunk 7 above)
Comment (artvalborgcp) [reply terraform/mynetwork.tf]: deleted

File: terraform/variables.tf (hunk 17 above)
Comment (artvalborgcp) [reply terraform/variables.tf]: made

File: terraform/variables.tf (hunk 18 above)
Comment (artvalborgcp) [reply terraform/variables.tf]: made

File: terraform/variables.tf (hunk 19 above)
Comment (artvalborgcp) [reply terraform/variables.tf]: done

File: terraform/variables.tf (hunk 16 above)
Comment (artvalborgcp) [reply terraform/variables.tf]: added

File: terraform/variables.tf (hunk 15 above)
Comment (artvalborgcp) [reply terraform/variables.tf]: added

File: terraform/variables.tf (hunk 13 above)
Comment (artvalborgcp) [reply terraform/variables.tf:181]: added

File: terraform/variables.tf (hunk 14 above)
Comment (artvalborgcp) [reply terraform/variables.tf:268]: added

File: terraform/outputs.tf [hunk 20]
```diff
@@ -0,0 +1,14 @@
+output "google_sql_database_instance_ip" {
+  value = google_sql_database_instance.sqlinstance.ip_address.0.ip_address
+}
+
+
+output "google_sql_database_instance_id" {
```
Comment (yevheniimykhailin) [inline terraform/outputs.tf:6]: add description

File: terraform/outputs.tf [hunk 21]
```diff
@@ -0,0 +1,14 @@
+output "google_sql_database_instance_ip" {
```
Comment (yevheniimykhailin) [inline terraform/outputs.tf:1]: add decription

File: terraform/mynetwork.tf [hunk 22]
```diff
@@ -0,0 +1,130 @@
...
+  timeout_sec   = var.timeout_backendservice
+  health_checks = [google_compute_http_health_check.mygcplbhealth.id]
+  backend {
+    group = google_compute_instance_group_manager.instance_group_manager.instance_group
+  }
+}
+
+
+
+resource "google_service_account" "isa" {
```
Comment (yevheniimykhailin) [inline terraform/mynetwork.tf:117]: should it go to iam.tf? if you split files - try to keep elements logically isolated.

File: 7-gce/config.py [hunk 23]
```diff
@@ -28,22 +28,22 @@
 # configure the respective settings for the one you choose below.
 # You do not have to configure the other data backends. If unsure, choose
 # 'datastore' as it does not require any additional configuration.
-DATA_BACKEND = 'datastore'
+DATA_BACKEND = os.environ.get('DATA_BACKEND')
```
Comment (yevheniimykhailin) [inline 7-gce/config.py]: os.environ.get('DATA_BACKEND', SOME_DEFAULT_VALUE)
the same below

File: 7-gce/bookshelf/tasks.py (hunk 1 above)
Comment (yevheniimykhailin) [reply 7-gce/bookshelf/tasks.py]: did you address that comment ^?

File: 7-gce/gce/startup-script.sh [hunk 24]
```diff
@@ -17,7 +17,29 @@
...
+echo  PROJECT_ID=$PROJECT_ID >> /etc/profile
+echo  region=$region >> /etc/profile
+echo  zone=$zone >> /etc/profile
+echo  DATA_BACKEND=$DATA_BACKEND >> /etc/profile
+echo  CLOUD_STORAGE_BUCKET=$CLOUD_STORAGE_BUCKET >> /etc/profile
+echo  CLOUDSQL_USER=$CLOUDSQL_USER >> /etc/profile
+echo  CLOUDSQL_PASSWORD=$CLOUDSQL_PASSWORD >> /etc/profile
+echo  CLOUDSQL_DATABASE=$CLOUDSQL_DATABASE >> /etc/profile
+echo  CLOUDSQL_CONNECTION_NAME=$CLOUDSQL_CONNECTION_NAME >> /etc/profile
+
```
Comment (yevheniimykhailin) [inline 7-gce/gce/startup-script.sh]: remove some blank lines

**Review by yevheniimykhailin — CHANGES_REQUESTED** [review]
All looks good, there are some tiny things to fix but all good.
Next PR should have modules.

File: 7-gce/bookshelf/tasks.py (hunk 1 above)
Comment (artvalborgcp) [reply 7-gce/bookshelf/tasks.py]: fixed

File: 7-gce/config.py (hunk 3 above)
Comment (artvalborgcp) [reply 7-gce/config.py]: fixed

File: terraform/outputs.tf (hunk 8 above)
Comment (artvalborgcp) [reply terraform/outputs.tf:10]: added

File: terraform/sql.tf (hunk 10 above)
Comment (artvalborgcp) [reply terraform/sql.tf]: made

File: terraform/sql.tf (hunk 11 above)
Comment (artvalborgcp) [reply terraform/sql.tf]: made

File: terraform/outputs.tf (hunk 20 above)
Comment (artvalborgcp) [reply terraform/outputs.tf:6]: added

File: terraform/outputs.tf (hunk 21 above)
Comment (artvalborgcp) [reply terraform/outputs.tf:1]: added

**Pull request merged**